pkg/publicshare/manager/cs3/cs3.go

```diff
@@ -0,0 +1,401 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package cs3
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/bcrypt"
+
+	user "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	link "github.com/cs3org/go-cs3apis/cs3/sharing/link/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	typespb "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/errtypes"
+	"github.com/cs3org/reva/v2/pkg/publicshare"
+	"github.com/cs3org/reva/v2/pkg/publicshare/manager/registry"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/indexer"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/indexer/option"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/metadata"
+	"github.com/cs3org/reva/v2/pkg/utils"
+)
+
+func init() {
+	registry.Register("cs3", NewDefault)
+}
+
+// Manager implements a publicshare manager using a cs3 storage backend
+type Manager struct {
+	storage          metadata.Storage
+	indexer          indexer.Indexer
+	passwordHashCost int
+
+	initialized bool
+}
+
+// PublicShareWithPassword represents a public share including its hashes password
+type PublicShareWithPassword struct {
+	PublicShare    *link.PublicShare `json:"public_share"`
+	HashedPassword string            `json:"password"`
+}
+
+type config struct {
+	GatewayAddr       string `mapstructure:"gateway_addr"`
+	ProviderAddr      string `mapstructure:"provider_addr"`
+	ServiceUserID     string `mapstructure:"service_user_id"`
+	ServiceUserIdp    string `mapstructure:"service_user_idp"`
+	MachineAuthAPIKey string `mapstructure:"machine_auth_apikey"`
+}
+
+// NewDefault returns a new manager instance with default dependencies
+func NewDefault(m map[string]interface{}) (publicshare.Manager, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		err = errors.Wrap(err, "error creating a new manager")
+		return nil, err
+	}
+
+	s, err := metadata.NewCS3Storage(c.GatewayAddr, c.ProviderAddr, c.ServiceUserID, c.ServiceUserIdp, c.MachineAuthAPIKey)
+	if err != nil {
+		return nil, err
+	}
+	indexer := indexer.CreateIndexer(s)
+
+	return New(s, indexer, bcrypt.DefaultCost)
+}
+
+// New returns a new manager instance
+func New(storage metadata.Storage, indexer indexer.Indexer, passwordHashCost int) (publicshare.Manager, error) {
+	return &Manager{
+		storage:          storage,
+		indexer:          indexer,
+		passwordHashCost: passwordHashCost,
+		initialized:      false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
```

Comment:
Would it make sense to lock? Then we do not need to make unnecessary calls to `Init`, `MakeDirIfNotExists` and `AddIndex`